### PR TITLE
[NCGenerics] Serialization for noncopyable generic types.

### PR DESCRIPTION
**Pull request by kavon**
This PR covers serialization and deserialization of ~Copyable types when NoncopyableGenerics is enabled. This means that you can now write:

```
public protocol P: ~Copyable {}
```

and emit Swift module containing that type, to be imported by another module.

resolves rdar://118947007&119531975

### CODE REVIEW

**Comment by kavon** [comment]
@swift-ci please test

**Comment by kavon** [comment]
https://github.com/apple/llvm-project/pull/7842
@swift-ci please smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci please smoke test

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -1447,6 +1457,40 @@ static unsigned getDepthOfType(Type ty) {
   return depth;
 }
 
+/// Recomputes which inverses must have been written for the given generic
+/// signature.
+static void reconstituteInverses(GenericSignature genericSig,
+                                 ArrayRef<Type> genericParams,
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp:1463]: these are not just generic parameters, they can also be dependent member types `Self.A` or a protocol, so perhaps typeParams or defaultedTypeParams is a better name.

File: lib/AST/ASTPrinter.cpp [hunk 2]
```diff
@@ -1558,10 +1602,20 @@ bestRequirementPrintLocation(ProtocolDecl *proto, const Requirement &req) {
 }
 
 void PrintAST::printInheritedFromRequirementSignature(ProtocolDecl *proto,
-                                                      Decl *attachingTo) {
+                                                      TypeDecl* attachingTo) {
+  Type attachedGP[1];
+  if (auto proto = dyn_cast<ProtocolDecl>(attachingTo))
+    attachedGP[0] = proto->getSelfInterfaceType();
+  else if (auto assoc = dyn_cast<AssociatedTypeDecl>(attachingTo))
+    attachedGP[0] = assoc->getDeclaredInterfaceType();
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp:1610]: this should be exhaustive otherwise attachedGP[0] is uninitiazed

File: lib/AST/Requirement.cpp [hunk 3]
```diff
@@ -260,57 +259,54 @@ InvertibleProtocolKind InverseRequirement::getKind() const {
...
 }
 
-
 void InverseRequirement::enumerateDefaultedParams(
-    TypeDecl *decl,
+    GenericContext *genericContext,
     SmallVectorImpl<Type> &result) {
 
   auto add = [&](Type t) {
     assert(t->is<GenericTypeParamType>() || t->is<DependentMemberType>());
```
Comment (slavapestov) [inline lib/AST/Requirement.cpp:267]: assert(t->isTypeParameter());

File: lib/Serialization/Deserialization.cpp [hunk 4]
```diff
@@ -6920,11 +6920,16 @@ DESERIALIZE_TYPE(GENERIC_TYPE_PARAM_TYPE)(
 
 Expected<Type> DESERIALIZE_TYPE(PROTOCOL_COMPOSITION_TYPE)(
     ModuleFile &MF, SmallVectorImpl<uint64_t> &scratch, StringRef blobData) {
-  bool hasExplicitAnyObject;
+  bool hasExplicitAnyObject, hasInverseCopyable, hasInverseEscapable;
```
Comment (slavapestov) [inline lib/Serialization/Deserialization.cpp:6923]: Maybe we should have an OptionSet for PCT flags at this point

**Comment by kavon** [comment]
@swift-ci please smoke test

**Comment by kavon** [comment]
@swift-ci please smoke test macOS platform

**Comment by kavon** [comment]
@swift-ci please smoke test

**Comment by kavon** [comment]
@swift-ci please smoke test

**Comment by kavon** [comment]
@swift-ci please smoke test linux platform

**Comment by kavon** [comment]
@swift-ci please smoke test linux platform

**Pull request merged**